### PR TITLE
When Jack is used, start / stop jack transport

### PR DESCRIPTION
**Pull request by djfm**
This allows lmms to function as a jack transport master.

Ideally there should be a switch to toggle between jack master / jack slave but this is already enough to do things like recording audio in ardour and have it sync'd with lmms easily.

### CODE REVIEW

**Comment by SecondFlight** [comment]
This appears to not be building on mingw, which I believe is Windows. Do you maybe need a preprocessor directive to prevent your logic from compiling in when Jack isn't present? This is my best guess without actually inspecting your code.

**Comment by SecondFlight** [comment]
Yep, Travis is only failing on Windows.

**Comment by SecondFlight** [comment]
Awesome, looks like the build is fixed. I'll let someone with more knowledge of JACK decide if this is good to merge.

**Comment by djfm** [comment]
As such the feature is pretty limited : it only starts and stops the jack transport when playback is started / stopped in LMMS. I think the next step would be to to also synchronize the play position when the song is stopped or the user skips to a random location in LMMS.

IMHO the feature as is already provides significant value because it makes recording audio with jack clients a lot easier, which is an important use case that is not handled by LMMS itself.

I'd like to go further and add support for the [jack_transport_reposition](http://jackaudio.org/api/group__TransportControl.html#ga2f371010358add3cbed8454bd0bd2ef8) API.

If anyone could share some LMMS knowledge about how to fill in a [jack_position_t](http://jackaudio.org/api/structjack__position__t.html) struct from LMMS timing info it would be very helpful. It seems to me like Song::setPlayPos would be a good hook to implement such feature, but I'm very unfamiliar with the LMMS codebase so any guidance on where it is most appropriate to put the jack sync code would be most welcome also.

**Comment by PhysSong** [comment]
> I'd like to go further and add support for the [`jack_transport_reposition`](http://jackaudio.org/api/group__TransportControl.html#ga2f371010358add3cbed8454bd0bd2ef8) API.

:+1:
> If anyone could share some LMMS knowledge about how to fill in a [`jack_position_t`](http://jackaudio.org/api/structjack__position__t.html) struct from LMMS timing info it would be very helpful.

I guess these functions will be useful:
https://github.com/LMMS/lmms/blob/2f19fa11c886374adcc74990e52d3ed0473fbf08/include/Song.h#L105-L108
https://github.com/LMMS/lmms/blob/2f19fa11c886374adcc74990e52d3ed0473fbf08/include/Song.h#L124-L140

**Comment by djfm** [comment]
Thanks a lot! I'm still struggling to make the mapping, is a "Tact" the same as a "Bar"? I can't seem to find a way to get the current bar.

**Comment by rghvdberg** [comment]
Tact is bar

**Comment by djfm** [comment]
Thanks, yeah me too but it's proving to be harder than I thought, care to help ? :)

I've got this piece of code together to fill in the jack position struct:

```C++
jack_position_t pos;
pos.valid = jack_position_bits_t::JackPositionBBT;
pos.bar = m_song->currentTact() + 1;
pos.beat = m_song->getBeat() + 1;
pos.tick = m_song->getBeatTicks() + 1;
pos.bar_start_tick = m_song->getTicks();
pos.beats_per_bar = m_song->getTimeSigModel().numeratorModel().value();
pos.beat_type = m_song->getTimeSigModel().denominatorModel().value();
pos.ticks_per_beat = m_song->getPlayPos().ticksPerBeat(m_song->getTimeSigModel());
pos.beats_per_minute = m_song->getTempo();
```

and then I do

```C++
jack_transport_reposition(m_client, &pos);
```

but Ardour will just go to around bar 1 and loop...

Can you spot something obviously wrong in the mapping code?

**Comment by rghvdberg** [comment]
Ok stupid noob question. Did you check if those functions actually return the correct bbt values?
By using either cout or qDebug()?

**Comment by djfm** [comment]
Yes I did, the values seem to make sense:

```
bar 3 beat 2 tick 13 ticks_per_beat 48 tempo 140
bar 4 beat 4 tick 13 ticks_per_beat 48 tempo 140
bar 5 beat 3 tick 37 ticks_per_beat 48 tempo 140
```

**Comment by djfm** [comment]
I'm getting closer, I had forgotten to set the "frame" field in the jack position object. The sync is now almost working, I can skip in lmms and it skips to what seems to be the correct position in ardour.

The **problem is** that I don't really know where to put this sync code. I think jack_transport_reposition should be called:
- when moving the playhead in lmms by clicking on the timeline or otherwise, but not when playing normally
- when the song is stopped or paused

At the moment I have the sync code in `SongEditor::updatePosition`, which has the side-effect of constantly stopping and restarting the jack transport, making synce'd clients useless.

**Comment by djfm** [comment]
As a workaround for not finding a suitable place to hook into to adjust the transport position, I'm only running the sync code when the song is paused. It's not ideal but usable.

**Comment by PhysSong** [comment]
I guess you may use [JACK timebase callback](http://jackaudio.org/api/group__TransportControl.html#ga5861cf9b84a0a3c3f6ff0f193ca3ec2c) for this purpose. Setting the callback notify JACK that LMMS is working as transport master.
I also think it'd be better if users can turn on/off transport setting, too.

**Comment by PhysSong** [comment]
@djfm Are you willing to continue this pull request? If not, some of us can continue working on this.

File: CMakeLists.txt [hunk 1]
```diff
@@ -380,7 +380,7 @@ IF(WANT_JACK)
 			SET(STATUS_JACK "OK (weak linking enabled)")
 			SET(JACK_INCLUDE_DIRS "")
 			# use dlsym instead
-			SET(JACK_LIBRARIES ${CMAKE_DL_LIBS})
+			SET(JACK_LIBRARIES ${CMAKE_DL_LIBS} jackserver)
```
[the remark this answers is not in the document]
Comment (tresf) [reply CMakeLists.txt]: It was failing the build as well.  Removed.

**Comment by PhysSong** [comment]
The new build failure is due to missing `jack_transport_reposition` in `weak_libjack.def`. It will be nice if someone could create a PR against upstream, or I can do that no one is going to.

**Comment by lukas-w** [comment]
> The new build failure is due to missing `jack_transport_reposition`

Can we use [`jack_transport_locate`](http://jackaudio.org/files/docs/html/group__TransportControl.html#gab3f52a42084aead87fd1ee75ed25b240) instead?

**Comment by firewall1110** [comment]
Was not able to compile:
```
/home/wb/build/lmms_build/check_jackd_ss/lmms/src/3rdparty/qt5-x11embed/src/X11EmbedContainer.cpp:40:10: fatal error: QtCore/private/qobject_p.h: No such file or directory
 #include <QtCore/private/qobject_p.h>
```

Is this PR actual?

Is it make sense to try  "insert this" against (for example) LMMS stable code base?

**Comment by PhysSong** [comment]
@firewall1110 It seems like you're missing Qt private headers for `qtbase`. Try searching packages for your distribution and install it.

**Comment by firewall1110** [comment]
quote:"_It will be nice if someone could create a PR against upstream, or I can do that no one is going to._"
But it was mentioned ~21 month ago ... 
May be this is done or in somebody plans or is not actual now. 

P.S.
_Start position:_ I had compiled stable and master + ~5 PR ... 
(1) To satisfy CMake **in this PR** I have added +3 packages
(2) CMake _download some code in configuration process_ and it seems that not find "one thing". So I "was forced" to -DWANT_CALF=OFF 
(3) It possible, that I should install some more packages, but CMake not recognised this (packages absense). 
Anyway this PR have problems with compilation (not only my building enviroppment)

This PR change only 6 files and time to manually apply this change to stable (this not download anything in configuration and "is stable" - not changing every week) may be the same as time to solve compilation problem

**Comment by PhysSong** [comment]
@firewall1110 Have you read the instruction for getting submodules?
https://github.com/LMMS/lmms/wiki/Accessing-git-repository
If you have any issues with compilation, please feel free to join and ask question on Discord.

**Comment by firewall1110** [comment]
I "solved" compilation problem by manual adapting code against master:
[pmaster.patch.gz](https://github.com/LMMS/lmms/files/5506489/pmaster.patch.gz)

I am an one step to "_create a PR against upstream_" ... 

But I do not see issue, that solves this code ...

**Comment by ycollet** [comment]
I "rebased" the changes on the current master here:
https://github.com/ycollet/lmms/tree/jcak_master
I don't know how to push this change on this pull request yet ...

**Comment by firewall1110** [comment]
It is good that @ycollet is about to work with this jackd integration with LMMS .
From my point of view more advanced jackd integration should be optional and not needed. 
   Plan may be :
(1) New PR should be created. 
(2) This PR closed.
(3) New PR reviewed and merged.

P.S.
I will not work in this direction.

**Comment by ycollet** [comment]
Thanks for your answer :)
For me a better jack integration (being able to use jack master / slave) should allow to use lmms with xjadeo.
The goal is to use lmms to write music in sync with a video.

**Comment by firewall1110** [comment]
LMMS with xjadeo now can be used (pmaster.patch.gz , and (I hope) Your "rebased" this PR version). 
(4) May be good "Ex.Sync." toggle in LMMS main window.
_Are You able to develop this Yourself?_

P.S.
I have only 1 (merged bugfix) PR and in seems as You "in wonder" with LMMS project management ...
("in wonder" not means "bad", but  - "I not understand")

**Comment by firewall1110** [comment]
This PR is not replaced yet ?
I am about to return ...

https://user-images.githubusercontent.com/57725851/122440537-535ba300-cfa5-11eb-93f0-422f5097ab8b.mp4

I will wait about 10 days . After I will realize my plan myself ...

**Comment by ycollet** [comment]
Excellent, I will just say GO !
Once your work is commited into LMMS, I have a lot of this to do with LMMS + Xjadeo !! :)

**Comment by firewall1110** [comment]
So I should not wait 10 days but do work it 10 days …
My previous plan is realized in draft:

https://user-images.githubusercontent.com/57725851/122562405-b56ee380-d04b-11eb-8cac-3aeabf2c160e.mp4

But:
<A>  legacy ugly solution: … dynamic_cast<AudioJack*> … 
(1) ExSync is possible not only using AudioJack (in theory) and it is possible to have more “slave target”. 
(2) AudioJack may be compiled in but not using  right now , especially than audio driver change without LMMS restart will be implemented. 
(3) jackd now may be used not only with  AudioJack but by jackd back-end in some other drivers – may be ExSync can be used too ;
But right now ExSync is possible  only  with  AudioJack and should be placed near and compiled with it, but not as  AudioJack  part … .
<B> ExSync is placed and implemented in the possible for me now way but not the good way:
I want to place button here (Is there something like this in QT):
exsync_place
<C> now  repositionTransport(pos) is called on every GUI synchronization , but should only as reaction on needed GUI events; for xJadeo “slave” current solution is enough , but for “audio slave” sound will be corrupted; 
I have to find connected slots for:
marked_points

**Comment by firewall1110** [comment]
Previous plan is realized in draft except place for button – I plan Slave/Duplex support:
lmms_jackd_image06_21
Next plan:
(1) Enable Slave/Duplex mode (message receiving from jackd).
(2) Work around LMMS reaction to user interface in Slave mode.
When critical user event happens, one of two thing should be happened:
- disconnect from jack-transport;
OR
- first aromatically change to Duplex mode.   
But may be allow strange variant , when LMMS will have “own life” to next jack-transport signal … But this will look like bug … 
(3) Work around potential infinite message loop in Duplex mode (make sure not send back received  messages). 
(4) Work around currentFrame(); wrong result if user work in PianoRoll (when SongEditor position is not updated)
(5) Place common part to Song.cpp , but Jackd part to AudiioJack.cpp. But type definitions leave in AudioJack.h (when will be more external synchronization possibilities than AudioJack, type definitions should be moved to another place).  
(6) Remove debugging lines , check build without JACK and prepare new PR that enables basic Jack Transport support and external synchronization experimental feature in draft.

**Comment by ycollet** [comment]
Thanks a lot for the update !

**Comment by firewall1110** [comment]
I just have opened PR #6066 so this should be closed ...

**Comment by PhysSong** [comment]
Closing in favor of #6066.

**Pull request closed without merging**